### DISCUSSION TITLE
What is leaked space and what can be done about it?

### DISCUSSION THREAD

**Discussion by sebveit** [openzfs/zfs#15159] [Q&A] [answered]
Dear devs,

I've encountered leaked space in one of my pools while trying to figure out what is wrong with that pool. (https://github.com/openzfs/zfs/issues/15150)
The documentation (man-pages) of zfs is very detailed, but I still don't know what impact leaked space has.

Quote from the pages (man zpoolprops):
> leaked:    Space not released while freeing due to corruption, now
		  permanently leaked into the pool.

It seems to me, that the space, which should be recycled, couldn't be logged/recorded.
Can that leaked space be reused/reclaimed in the future with a possible scrub?
Is it "lost" forever in the deeps of the vdev?

Thanks in advance!

**Comment by rincebrain** [comment, marked as answer]
Leaked means "you have lost it forever, we decided to mark it as used and never think about it again because you enabled a tunable that told us to do this".

**Comment by sebveit** [comment]
Thank you for your fast reply! I guess I got it.

Let me rephrase it for someone who isn't that informed about ZFS: If you stumble upon a zpool (e.g. `zpool get all name-of-the-pool`), which has the property `leaked` with a non-zero value, that value is the amount of space that is ignored by ZFS on that particular zpool. No scrubbing or resilvering that zpool will make that space useable in the future. The zpool will function normally and no errors or events (`zpool events name-of-the-pool`) will be reported regarding that leaked space.

A theoretical utility (e.g. the infamous zfs-fsck) could return this lost space into the accounting of ZFS, since that leaked space is known to the zpool (e.g. logged where it is and how much it occupies). Any data/information in this space can not be interpreted/recovered and would simply be filled with zeros and reintegrated as free space.

Is that right or am I missing something? Thanks!

**Reply by rincebrain** [reply]
It would depend on why the space was leaked whether a purely theoretical utility could recover it or not.

In general, the reason it was leaked in the first place would still preclude you doing that, I would expect, but I don't have an exhaustive list in my head of the reasons that it might happen.

Also in general, "telling it to just leak the space is probably not your best option."

**Reply by sebveit** [reply]
I shot myself in the foot while trying to fix the problem I'm describing in https://github.com/openzfs/zfs/issues/15150.
But the amount of leaked space is minimal (128 megabyte) ... so as long as the pool isn't affected ...
 
Ok, thanks for your insight. I mark your answer and I guess this discussion can be closed.
I will post updates in the mentioned issue I have opened.
